OAK-11861 - segment store initialization should run in parallel

jsedding opened the pull request.

jsedding committed:
OAK-11861 - segment store initialization should run in parallel

joerghoh commented on oak-segment-tar/src/main/java/org/apache/jackrabbit/oak/segment/file/tar/TarFiles.java:
```diff
@@ -411,18 +413,28 @@ public void init() throws IOException {
+                        }
+                    } catch (IOException e) {
+                        log.warn("Unable to open TAR file: {}", map.get(index), e);
+                        throw new UncheckedIOException(e);
```
how is the exception handling for this UncheckedIOException? And would it possible to pass some more information (for example the name of the file (it's logged before).

jsedding replied:
I have added handling to unwrap the original `IOException`.

reschke commented on oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/concurrent/package-info.java:
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
-@Version("1.0.1")
+@Version("1.1.0")
```
Please do not put new methods into public APIs unless they are really needed outside Oak. In the past we created new packages for that.

See https://github.com/apache/jackrabbit-oak/pull/2457 which happens to do that.

jsedding replied:
Moved to `forkjoin` package and marked it as internal API.

reschke commented on oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/concurrent/ForkJoinUtils.java:
```diff
@@ -0,0 +1,120 @@
+/*
```
There should be unit test coverage inside oak-commons.

jsedding replied:
Added a unit-test.

jsedding committed:
OAK-11861 - segment store initialization should run in parallel

- move ForkJoinUtils to new, internal package
- add unit test for ForkJoinUtils

jsedding commented:
@reschke I have addressed your comments in commit 7290b84. Unless there are any other issues, I will merge the PR once the build completes successfully.

jsedding committed:
OAK-11861 - segment store initialization should run in parallel

- export new forkjoin package

jsedding commented:
@reschke can you please review again and confirm that my changes address your concern? Thanks!

reschke reviewed (approved):
I did not check the actual change. My high-level concerns were adressed.

That said, please have a look ath the remaining Sonar warnings (https://sonarcloud.io/project/issues?id=org.apache.jackrabbit%3Ajackrabbit-oak&pullRequest=2454&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)

reschke commented:
late comment: for consistency with other packages in commons (like "pio"), I'd make this "pconcurrent".
